muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,771 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use bytes::Bytes;
+use crate::{MplexConfig, MaxBufferBehaviour};
+use crate::codec::{Codec, Frame, LocalStreamId, RemoteStreamId};
+use log::{debug, trace};
+use fnv::FnvHashMap;
+use futures::{prelude::*, ready, stream::Fuse};
+use futures::task::{ArcWake, waker_ref, WakerRef};
+use futures_codec::Framed;
+use parking_lot::Mutex;
+use std::collections::{VecDeque, hash_map::Entry};
+use std::{cmp, io, mem, sync::Arc, task::{Context, Poll, Waker}};
+
+pub use std::io::{Result, Error, ErrorKind};
+
+/// A multiplexed I/O stream.
+pub struct Multiplexed<C> {
+    /// The current operating status.
+    status: Status,
+    /// The underlying I/O stream.
+    io: Fuse<Framed<C, Codec>>,
+    /// The configuration.
+    config: MplexConfig,
+    /// Buffer of received frames that have not yet been consumed.
+    buffer: Vec<Frame<RemoteStreamId>>,
+    /// Whether a flush is pending before reading frames can proceed.
+    pending_flush: bool,
```

Comment:
Do I understand that this is only `true` when an `Open` frame is still to be flushed? If so, how about renaming it to `pending_open_frame_flush`. I was a bit surprised `poll_read_frame` yields to flushes in general, even though it only does for `Open` frames, which makes sense.

Reply:
> Do I understand that this is only true when an Open frame is still to be flushed?

Yes, when at least one new outbound stream has been opened (i.e. an `Open` frame was sent) but not yet flushed before trying to read. Would you be ok with the name `pending_flush_open`?

Reply:
`pending_flush_open` rolls off the tongue much better, thanks!

Reply:
Done in https://github.com/libp2p/rust-libp2p/pull/1769/commits/37c464c7a950ccbc9e7aba5380c02887af41e79a.